component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java

```diff
@@ -0,0 +1,881 @@
+
+public class BackupCodeAuthenticator extends AbstractApplicationAuthenticator
+        implements LocalApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(BackupCodeAuthenticator.class);
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+
+        if (context.isLogoutRequest()) {
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        } else if (request.getParameter(BACKUP_CODE) == null) {
+            initiateAuthenticationRequest(request, response, context);
+            if (context.getProperty(AUTHENTICATION)
+                    .equals(BACKUP_CODE_AUTHENTICATOR_NAME)) {
+                return AuthenticatorFlowStatus.INCOMPLETE;
+            } else {
+                return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+            }
+        } else {
+            return super.process(request, response, context);
+        }
+    }
+
+    @Override
+    public boolean canHandle(HttpServletRequest httpServletRequest) {
+
+        String token = httpServletRequest.getParameter(BACKUP_CODE);
+        String action = httpServletRequest.getParameter(BackupCodeAuthenticatorConstants.SEND_TOKEN);
+        String enableBackupCode = httpServletRequest.getParameter(BackupCodeAuthenticatorConstants.ENABLE_BACKUP_CODE);
+        return (token != null || action != null || enableBackupCode != null);
+    }
+
+    @Override
+    public String getContextIdentifier(HttpServletRequest httpServletRequest) {
+
+        return httpServletRequest.getParameter("sessionDataKey");
+    }
+
+    @Override
+    public String getName() {
+
+        return BackupCodeAuthenticatorConstants.BACKUP_CODE_AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    public String getFriendlyName() {
+
+        return BackupCodeAuthenticatorConstants.BACKUP_CODE_AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String username = null;
+        String tenantDomain = context.getTenantDomain();
+        context.setProperty(AUTHENTICATION,
+                BackupCodeAuthenticatorConstants.BACKUP_CODE_AUTHENTICATOR_NAME);
+        if (!tenantDomain.equals(BackupCodeAuthenticatorConstants.SUPER_TENANT_DOMAIN)) {
+            IdentityHelperUtil
+                    .loadApplicationAuthenticationXMLFromRegistry(context, getName(), tenantDomain);
+        }
+        AuthenticatedUser authenticatedUserFromContext = BackupCodeUtil.getAuthenticatedUser(context);
```

Comment:
Find whether there is already an utility method to do this.

Reply:
There are no public methods for this